### PR TITLE
Train mem usage and other improvements 

### PR DESCRIPTION
**Pull request by xaedes**
This PR improves the memory requirements of training by **removing unnecessary tensors from adam optimizer** and with a (manual) **gradient checkpointing** training function, which reduces memory overhead from O(n_layer) to O(sqrt(n_layer)) as explained in readme of https://github.com/cybertronai/gradient-checkpointing

Other changes:
- Adam(W) optimizer now supports **gradient clipping** which improves training convergence
- optimizers can now be passed a callback, which is called before each optimization iteration. This can be used to set a learning rate based on custom learning schedule and **change sample batches between each iteration**, which improves training convergence and run time by avoiding the overhead of restarting the optimization loop to update learning rate and sample batches
- fixed some issues in cross entropy loss and improved numerical stability of backward pass with a simplified computation (as in other frameworks), assuming target probability vector sums to one for each batch 
- cross entropy loss now returns the mean loss over all batches, which helps keeping the gradients in a sane range and decouples the gradients from the batch size
- change AdamW decay parameter to work like the torch AdamW decay param, and change default AdamW weight decay parameter defined in ggml to 0.0, making Adam default instead of AdamW
- add conditional compilation for F16 exp in flash attention and cross entropy loss, improving gradient quality 
- ggml : update ggml_rms_norm_back with configurable eps
- ggml : add function `ggml_build_backward_expand` to avoid stack overflows with large maximum number of nodes: `GGML_API void ggml_build_backward_expand(struct ggml_context * ctx, struct ggml_cgraph * gf, struct ggml_cgraph * gb, bool keep);`

- expose more optimizer parameters as training parameters
- change sampling parameters for prediction after training to defaults of common.h and clarify what is context for prediction and what are generated tokens

### CODE REVIEW

**Comment by ggerganov** [comment]
Very interesting work!

There is a lot of manual work - wonder what we can do to improve this. Have you looked at the new memory allocator work by @slaren? It might be a way to avoid the use of scratch

**Comment by xaedes** [comment]
Removing the scratch buffers would certainly help! 
When this works an automatic graph rewrite for gradient checkpointing could be possible.
I did not see the allocator yet, will look into it.

**Comment by xaedes** [comment]
@slaren 

> there isn't any way to specify what tensors must not be overwritten

Since the backward pass also just builds a computational graph with proper dependencies between the nodes, I think it could work as it is: by just calling the allocator on the backward graph instead of on the forward graph.

**Comment by howard0su** [comment]
We may need to do DAG analysis to figure out the dependency. With the new allocator, the result can guide the allocation and free.

**Comment by xaedes** [comment]
After some fiddling I got a version working with unified function for building compute forward and backward graphs for training using the memory allocator and an (optional) automatic checkpointing graph rewrite function.

So far in my tests the handwritten train function using scratch buffers requires 30% less memory than the automatic version.
I also found differences in the resulting gradients and optimized parameters, hinting at some (probably tiny) bug in the handwritten version. 

I guess the function using allocator could also work with ~30% less memory by optimizing the order of operations, the memory allocation strategy or both.

But this gets too much for this pull request, so I think I will wrap it up and push the modifications, replacing lots of handwritten code with the version using automatic memory allocation and gradient checkpointing.

LORA finetuning implementation will get much easier when using the automatic mem alloc and grad checkpointing. We can always improve on those automatic mechanism later on.

**Comment by xaedes** [comment]
Didn't want to make just one big commit, so I separated it in multiple commits. 
I messed something up during this commiting process, will need to find the bug. 
Will comment here when it is fixed.

**Comment by xaedes** [comment]
The last commits fixed the bug. Recompiling everything also helped squash some weird issues I encountered ^^. 

The resulting gradients and parameters are the same (tested using checksums of tensor contents) for each configuration of the bool options `use_checkpointing` & `use_alloc`. 

Threading or not also does not influence the gradients and resulting parameters. 
But the reported loss value differs due to different order of summation during threading, more threads resulting in more accurate reported loss value. This has zero influence on the gradients and model parameters though.

As mentioned before gradient checkpointing currently brings no actual benefit. Maximum size reported by allocator is nearly the same as when not using it: only saving 0.136% on a 16 layer model, at the cost of one additional forward pass.

**Comment by xaedes** [comment]
Just reversing the order of the last for-loop in `ggml_build_backward_expand` brought the memory improvements due to gradient checkpointing back to life.

With this loop order gradient checkpointing with allocator on 16 layer model saves 13% memory; 2 layer memory it still saves 3-4% memory.

The computation results are the same.

**Comment by ggerganov** [comment]
After finishing with #2398, will look into merging this

**Comment by xaedes** [comment]
> After finishing with #2398, will look into merging this

When 2398 is merged I would first change this PR to use GGUF for the checkpointing files. 
This PR here updates the checkpointing file format anyway. Already looked into the state of GGUF to use it here, but since it was not yet merged I could only update my own file format. Directly using GGUF could eliminate the checkpointing file format which I would greatly prefer.

**Comment by xaedes** [comment]
Working on the GGUF stuff for saving models and loading/saving checkpoints.

**Comment by xaedes** [comment]
Since the gguf constants are only in `llama.cpp` I can't access them, so I have to replicate them in `train-text-from-scratch.cpp`

**Comment by ggerganov** [comment]
Yeah, I was wondering about that. For now lets replicate. Later we might expose them in the API

**Comment by xaedes** [comment]
@ggerganov @philpax

Hm... for saving the number of optimization parameters (in checkpointing file) I need a 64 bit unsigned integer, because this can actually exceed 2^32, e.g. in a >4B model. 

But there are only GGUF types with at most 32 bit, or arrays & strings.

I mean I could split the 64bit-integer into two uint32_t words, and/or store them as array or string. 
That feels a bit hackish, but I guess it is a bit late for adding more GGUF types. 
So I will probably just do it the hackish way, splitting and storing them as array values.
A similar issue will arise, when you want to store a double (float64) value. 

The parameter count is necessary for the optimizer state tensors.

There comes the next problem: These are 1D-tensors with the size of the parameter count.
GGML tensors store number of elements for each dimension with 64 bit precision, but GGUF will store `ne` with only 32 bit precision.

https://github.com/ggerganov/llama.cpp/blob/0d3094f0c742ce61f84feb6e4f0b59beee6194d7/ggml.h#L447-L452

https://github.com/ggerganov/llama.cpp/blob/0d3094f0c742ce61f84feb6e4f0b59beee6194d7/ggml.c#L19465-L19469

The parameter count may not be divisible to reshape the elements among multiple the other dimensions.
For this I will probably have to resize the tensors up to the next reshapeable parameter count and then reshape them for storage in GGUF.

**Comment by ggerganov** [comment]
Hm, this might be a real issue. Need to log off now and will think some more about this tomorrow, but on first glance we need to change `ne` in `gguf_tensor_info` to `int64_t` and in general support `int64_t` types.

**Comment by philpax** [comment]
Erk. That's awkward. We can do a GGUF v2 (the format supports this), but it would suck to have to do a break so soon, especially for the users who just downloaded quantized models. It could be done in the future, or llama.cpp could convert v1 to v2 (should be lossless).

**Comment by xaedes** [comment]
Okay, I will wait then until we know how to proceed and also call it day for today.

**Comment by ggerganov** [comment]
Let's change `ne` to int64_t in `gguf_tensor_info` and add support for read/write 64-bit ints.
I'll add support for GGUFv1 by reading 32-bit `ne`, so it remains compatible

**Comment by xaedes** [comment]
I think any changes to GGUF itself should be done outside of this PR.
I will then merge it into this PR from master.

**Comment by ggerganov** [comment]
Ok, I'll do this today or tomorrow

**Comment by xaedes** [comment]
Implemented GGUF checkpoint files and writing of GGUF models.
Use `convert-train-checkpoint-to-gguf.py` to convert old checkpoint files into GGUF format.

Fixes a bug that resulted in bad gradients during training.

This also fixes a memory-corruption bug when attempting to write any GGUF file from cpp:

> ctx->kv and ctx->infos was reallocated using not-aligned realloc, but freed with aligned free.
to fix this a GGML_ALIGNED_REALLOC was added, but there is no posix_memalign_realloc function.
so on non-windows and non-mingw32 platforms we fall back to aligned malloc, followed by copying
and freeing the old data.

All the prediction code is removed from train-text-from-scratch to reduce code duplication with main.
Use `main` for prediction instead.

File: ggml.c [hunk 1]
```diff
@@ -20065,7 +20077,7 @@ static int gguf_get_or_add_key(struct gguf_context * ctx, const char * key) {
 
     const int n_kv = gguf_get_n_kv(ctx);
 
-    ctx->kv = realloc(ctx->kv, (n_kv + 1) * sizeof(struct gguf_kv));
+    ctx->kv = GGML_ALIGNED_REALLOC(ctx->kv, n_kv * sizeof(struct gguf_kv), (n_kv + 1) * sizeof(struct gguf_kv));
```
Comment (ggerganov) [inline ggml.c:20080]: I initially chose to use `GGML_ALIGNED_ALLOC` for the `kv` and `infos` for no good reason but simply because I wanted to keep the pattern of not using `malloc` anywhere in `ggml.c`. Eventually, I started needing to allocate arrays and string which no longer makes sense to have aligned alloc so obviously I broke the pattern anyway.

Therefore, I think it is better to just switch to using `malloc` in `gguf` and avoid the new `GGML_ALIGNED_REALLOC` macro. Sorry about this complication.

**Review by ggerganov — APPROVED** [review]
Brb, training a tiny shakespeare llama 😄
